### DISCUSSION TITLE
Any way to get Ccache hits with --coverage and unique paths?

### DISCUSSION THREAD

**Discussion by fuzzewuzze** [ccache/ccache#1483] [Questions & answers]
So part of our CI has builds go into folders named after the  SHA they came from as 1 build machine maybe building multiple  things at the same time.  
But because it seems like GCC puts an absolute path in when using --coverage we basically get no hits.
I've seen some comments about using -fprofile-dir or such, but i have had no luck or I am just doing it wrong.

Does anyone have any advice on if you can use --coverage in unique output folders and still have GCOV/GCOVR and cache hits?
Everything below the SHA folder is the repo root and would match, i almost need a way like CTCache has to strip part of the path out?

**Comment by jrosdahl** [comment]
There is currently no way to do this. The reason is that (as you already noted) GCC includes the current working directory (CWD) in the object file as part of the path to the corresponding `.gcda` file to create at runtime, so ccache includes the CWD in the input hash. It would be possible to make ccache skip including the CWD, but then the object file will be broken if building from another directory since the `.gcda` will be created in the wrong directory (or perhaps not at all).

One solution is to build in a separate filesystem namespace, for instance using some container solution, so that each build runs in a virtual build directory with the same path.

Another solution is of course to not use `--coverage` since the "program flow arcs" information (from the implicit `-fprofile-arcs` option) cannot be used anyway when building with ccache. But perhaps you're only interested in the `-ftest-coverage` part of `--coverage`? If so, use `-ftest-coverage` instead of `--coverage` in combination with `gcno_cwd` sloppiness.

**Comment by fuzzewuzze** [comment]
Got it, i think it'd be nice for there to be a way to ignore the CWD for those of us that know what we're doing and are always building from the same location in a CI pipeline for example

**Reply by jrosdahl** [reply]
> always building from the same location

Hmm, I guess you forgot a "not" there since you said that the directories are uniquely named ("_after the SHA they came_")?

>  i think it'd be nice for there to be a way to ignore the CWD

If we add an option to not include the CWD in the input hash then the object files will include the wrong directory for the `.gcda` file. That is, the `.gcda` file will not be created in the correct place when running the program. But since there actually is support for [relocating those files at runtime](https://gcc.gnu.org/onlinedocs/gcc/gcov/data-file-relocation-to-support-cross-profiling.html) I agree that such an option would make sense.
